src/doc/trpl/concurrency.md

```diff
@@ -0,0 +1,343 @@
+in our concurrent programs.
+
+As an example, here is a Rust program that would have a data race in many
+languages. It will not compile:
+
+```ignore
+use std::thread::Thread;
+use std::io::timer;
+use std::time::Duration;
+
+fn main() {
+    let mut data = vec![1u32, 2, 3];
+
+    for i in 0 .. 2 {
+        Thread::spawn(move || {
+            data[i] += 1;
+        });
+    }
+
+    timer::sleep(Duration::milliseconds(50));
+}
+```
+
+This gives us an error:
+
+```text
+12:17 error: capture of moved value: `data`
+        data[i] += 1;
+        ^~~~
+```
+
+In this case, we know that our code _should_ be safe, but Rust isn't sure. And
+it's actually not safe: if we had a reference to `data` in each thread, and the
+thread takes ownership of the reference, we have three owners! That's bad. We
+can fix this by using the `Arc<T>` type, which provides an atmoic reference
+count. The 'atomic' part means that it's safe to share across threads.
+
+Here's the second version of our code. It still doesn't work, but for a
+different reason:
+
+```ignore
+use std::thread::Thread;
+use std::io::timer;
+use std::time::Duration;
+use std::sync::Mutex;
+
+fn main() {
+    let mut data = Mutex::new(vec![1u32, 2, 3]);
+
+    for i in 0 .. 2 {
+        let data = data.lock().unwrap();
+        Thread::spawn(move || {
+            data[i] += 1;
+        });
+    }
+
+    timer::sleep(Duration::milliseconds(50));
+}
+```
+
+Here's the error:
+
+```text
+error: the trait `core::marker::Send` is not implemented for the type `core::marker::NoSend`
+```
+
+You see, [`Mutex`](std/sync/struct.Mutex.html) has a
+[`lock`](http://doc.rust-lang.org/nightly/std/sync/struct.Mutex.html#method.lock)
+method which has this siganture:
+
+```ignore
+fn lock(&self) -> LockResult<MutexGuard<T>>
+```
+
+If we [look at the code for MutexGuard](https://github.com/rust-lang/rust/blob/896cb36ecab3eaeb7f101087e030e43771eca5ca/src/libstd/sync/mutex.rs#L163-L170), we'll see
+this:
+
+```ignore
+__marker: marker::NoSend,
+```
+
+Because our guard is `NoSend`, it's not `Send`. Which means we can't actually
+transfer the guard across task boundaries, which gives us our error.
+
+The [`Arc<T>`](std/sync/struct.Arc.html) type is the answer. An `Arc<T>` allows
+us to reference count some data, which will allow our code to work. Here's the
+working version:
+
+```
+use std::sync::{Arc, Mutex};
+use std::thread::Thread;
+use std::io::timer;
+use std::time::Duration;
+
+fn main() {
+    let data = Arc::new(Mutex::new(0u32));
+
+    for _ in range(0, 10) {
+        let data = data.clone();
+        Thread::spawn(move || {
+            let mut data = data.lock().unwrap();
+            *data += 1;
+        });
+    }
+
+    timer::sleep(Duration::milliseconds(50));
+}
+```
+
+We now call `clone()` on our `Arc`, which increases the reference count. This
+reference is then moved into the new thread, which calls the `lock()` and
```

Comment:
I feel like readers are likely to be pretty lost without more explanation of the mutex API here...